### DISCUSSION TITLE
Segment grain boundaries into one unique segment

### DISCUSSION THREAD

**Discussion by eugenianietov** [mtex-toolbox/mtex#1884] [Ask Anything]
Hi!

I'm working with MTEX to get some data about the grain boundaries in a synthetically created EBSD, with every pixel containing the three Euler angles. Since it is a synthetic microstructure, grain boundaries are straight, and crystal orientations are homogeneous within the grains. However, when I perform the grain segmentation followed by smoothing with the smooth command, I still get several segments per grain boundary (which is logical since my microstructure is composed of square pixels). I tried to impose many iterations, like 1000, to the smooth command, but the grain boundary is still divided into smaller segments. Is there any way to force every grain boundary to be composed of just 1 single segment?

Thanks a lot in advance!

Eugenia

**Comment by eugenianietov** [comment]
Digging into the grainBoundary documentation, I found the function 'selectByGrainId'. I provided the list of grainId pairs, and I got 243 gBSub instead of the 4160 gB that I had before. Anyway, is this the purpose of the function, and is it doing what I intend? 

Thanks again!

**Comment by ralfHielscher** [comment]
May I ask why you want to reduce the number of boundary segments? 

Ralf.

**Reply by eugenianietov** [reply]
I want to reduce the number of boundary segments because I need to retrieve grain boundary data, such as misorientation, length, angle with the loading axis, ..., and it's way easier if I only have one GB segment per grain boundary! In fact, right now I'm looking for adjacent grain boundaries, and for that, it's more convenient to have two vertices per GB.

**Comment by eugenianietov** [comment]
> May I ask why you want to reduce the number of boundary segments?
> 
> Ralf.

Is there any way to do this then? Thanks a lot!

Eugenia

**Reply by ralfHielscher** [reply]
This is not so easy, I could implement a function that averages boundary segments with all their properties according to a specific criterion. However, it makes much more sense to consider the grain boundary segments individually. This  way you can also say something about variance of the properties you are interested in. 
I'm sure we can help you to analyse your grain boundary data. Such explain which kind of correlation you are interested in. 

All the best,
Ralf.

**Comment by vtvivian** [comment]
Hi Eugenia, 

I am also trying to figure this out! 

Possibly the approach suggested in https://github.com/mtex-toolbox/mtex/discussions/1492 maybe could work for you?
(Especially if your synthetic EBSD map is created by Voronoi tesselation, this should work exactly, as Delaunay triangulation is the dual graph of the Voronoi tesselation, i.e. they are 'inverted' versions of each other.) 
> Triple points have the properties x and y, possibly doing a Delauny triangulation on those and discarding those edges which should not dissect a grain (= same grainId on both sides) might be an idea.
> Cheers,
> Rüdiger

I don't know how to reconstruct the calculated line segments back into gb segments though - so you may have to create some workarounds to get the data you're interested in.

Vivian

**Comment by vtvivian** [comment]
Hi Eugenia, 

Does something like this, which joins the triple juctions with straight lines, work for your case? 
![data1_triplePointNetwork](https://github.com/mtex-toolbox/mtex/assets/27768054/d424a91e-3ab1-4014-bc95-ae4de137b5e7)
It doesn't work perfectly for a twinned microstructure, but would be better if your there are no isolated island grains and the boundaries don't split and rejoin. I've plotted these as matlab lines, but it should be possible to turn these into @grainBoundary segments if you can figure out how to use the grainBoundary constructor function. 

Vivian

<code>
% load demo dataset
mtexdata twins;
data1_ebsd = gridify(ebsd);
%
% this data1_grains is different to the one in demodata.m because the
% 'removeQuadruplePoints' flag is turned on
[data1_grains,data1_ebsd('indexed').grainId] = calcGrains(data1_ebsd('indexed'),'angle',10*degree,'removeQuadruplePoints');
%
%% find triple points - interior map points only
gB = data1_grains.boundary; 
tP = gB.triplePoints;
tPGrainList = tP.grainId; % list of grainIds per triple point
%make a helper variable to help us keep track of grainIds later on
tPIdList = repmat([1:length(tP)]',1,3);
%
%
%% identify gb segments on map edge
% only consider gb segments that are not part of the map edge
gB2 = gB('indexed','indexed');
%
% this is the map edge
[xmin, xmax, ymin, ymax] = extent(data1_ebsd);
% find gb segments which have a midpoint within 1/2 step of edge
% ebsd step size = data1_ebsd.dx or .dy 
% eps*1000 to get rid of rounding errors
%
edgeGrainId{1} = gB2(abs(gB2.midPoint.x-xmin) < data1_ebsd.dx/2-eps*1000).grainId; % xmin edge
edgeXY{1} = gB2(abs(gB2.midPoint.x-xmin) < data1_ebsd.dx/2-eps*1000).midPoint.xy; % coordinates
%
edgeGrainId{2} = gB2(abs(gB2.midPoint.x-xmax) < data1_ebsd.dx/2-eps*1000).grainId;  % xmax edge
edgeXY{2} = gB2(abs(gB2.midPoint.x-xmax) < data1_ebsd.dx/2-eps*1000).midPoint.xy; % coordinates
%
edgeGrainId{3} = gB2(abs(gB2.midPoint.y-ymin) < data1_ebsd.dy/2-eps*1000).grainId; % ymin edge
edgeXY{3} = gB2(abs(gB2.midPoint.y-ymin) < data1_ebsd.dy/2-eps*1000).midPoint.xy; % coordinates
%
edgeGrainId{4} = gB2(abs(gB2.midPoint.y-ymax) < data1_ebsd.dy/2-eps*1000).grainId; % ymax edge
edgeXY{4} = gB2(abs(gB2.midPoint.y-ymax) < data1_ebsd.dy/2-eps*1000).midPoint.xy; % coordinates
%
edgeGrainId=sort(cat(1,edgeGrainId{:}),2); %gId pairs
edgeXY=cat(1,edgeXY{:}); %xy positions
%
%% now find pairs of triple points that contain the same grainId pair
%
% tP.grainId % triplets of grainIds meeting at a triple point
% gB(tP.boundaryId).grainId %list of gb pairs associated w triple point segments
%
findPairs1 = @(x) sort(nchoosek(x, 2),2);
findPairs2 = @(x) reshape(permute(x,[2 1 3]),2,[]).';
%
for t=1:length(tP)
    % gId of neighbour grain pairs
    gIdPairs1(:,:,t) = findPairs1(tPGrainList(t,:)); 
    % tPIds corresponding to each neighbour grain pair
    tPIdPairs1(:,:,t) = findPairs1(tPIdList(t,:)); 
end
%
%reshape to find repeated rows later
gIdPairs2 = findPairs2(gIdPairs1);
tPIdPairs2 = findPairs2(tPIdPairs1);
%
%
% find pairs of tPId with the same grain boundary segment
[~,~,ib]=unique(gIdPairs2,'stable','rows');
%repeated values in ib correspond to repeated pairs of gId, i.e. same gb
%
%% method 1 - this misses out some gbs
% %run unique(ib) again to find the other occurences of each value
% [~,iaA,~]=unique(ib,'first');
% [~,iaB,~]=unique(ib,'last');
% 
% %now draw a boundary between data1_grains(tPIdPairs2(C(ix,1))) and data1_grains(tPIdPairs2(C(ix,2))) 
% % by connecting triple points  tPIdPairs2(iaA(ix),1) and tPIdPairs2(iaB(ix),1)
% % this method ignores gBs that connect to the map edge
% figure; plot(data1_grains.boundary); hold on;
% plot(tP,'color','b','linewidth',2)
% for ix=1:length(iaA)
%     tPA = tPIdPairs2(iaA(ix),1);
%     tPB = tPIdPairs2(iaB(ix),1);
%     % discard if these are the same point
%     if tPA==tPB
%         continue
%     end
%     gbLine{ix} = line([tP(tPA).x, tP(tPB).x].',[tP(tPA).y, tP(tPB).y].');
% end
%
%% method 2 - this connects too many boundaries
% most of the time we want to find pairs of triple points which share a
% grainId pair
% sometimes multiple pairings exist - e.g. if a gB loops and reattaches to
% the same gb segment. Find and connect all of them -- some of them are
% redundant.
for ix = 1:length(ib)
    v{ix}=find(ib==ib(ix)); 
    if numel(v{ix})>1
        %this gb segment connects to some other triple point(s)
        tPPairs{ix} = nchoosek(v{ix},2); %some of these boundaries don't exist
        tPEdgeA{ix}=[];
        tPEdgeB{ix}=[];
    else 
        tPPairs{ix} = [];
        %this gb segment connects to the edge of the map
        % find the gId pair on the map edge corresponding to grain pair gIdPairs2(ib(ix),:)
        [~,locB]=ismember(gIdPairs2(ix,:), edgeGrainId, 'rows');
        if locB>0
            % find the coordinates of the edge that connects to this tP
            tPEdgeA{ix} = [tP(tPIdPairs2(ix,1)).x tP(tPIdPairs2(ix,1)).y];
            tPEdgeB{ix} = edgeXY(locB,:);
        else
            tPEdgeA{ix} =[];
            tPEdgeB{ix}=[];
        end
    end
end
tPPairs=tPPairs(~cellfun('isempty',tPPairs));
tPEdgeA=tPEdgeA(~cellfun('isempty',tPEdgeA));
tPEdgeB=tPEdgeB(~cellfun('isempty',tPEdgeB));
%
%now find the coordinates of triple point pairs in tPPairs and connect them
% this method ignores gBs that connect to the map edge
tPPairs_all = cat(1,tPPairs{:});
tPEdgeA_all = cat(1,tPEdgeA{:});
tPEdgeB_all = cat(1,tPEdgeB{:});
tPA = tPIdPairs2(tPPairs_all(:,1),1);
tPB = tPIdPairs2(tPPairs_all(:,2),1);
%
%%
%
figure; 
plot(data1_grains, data1_grains.id); hold on; mtexColorbar;
plot(data1_grains.boundary,'color','k','linewidth',1);
gbLines1 = line([tP(tPA).x, tP(tPB).x].',[tP(tPA).y, tP(tPB).y].','linewidth',2,'color','c');
gbLines2 = line([tPEdgeA_all(:,1), tPEdgeB_all(:,1)].',[tPEdgeA_all(:,2), tPEdgeB_all(:,2)].','linewidth',2,'color','m');
plot(tP,'color','b','linewidth',1);
scatter(edgeXY(:,1),edgeXY(:,2),'r','linewidth',1);
%
savefig('data1_triplePointNetwork.fig');
export_fig('data1_triplePointNetwork.png');
</code>
